metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -397,8 +437,35 @@ def _ingest_events(
     ) -> None:
         """Read log and store events in usage_state."""
         num_aggregated = 0
+        num_generated = 0
         for audit_event in events:
             try:
+                # Note for View Usage:
+                # If Query Event references a view, bigquery audit logs do not contain Read Event for view
+                # in its audit logs, but only for it base tables. To extract usage for views, we parse the
+                # sql query to find bigquery tables and views read in the query and generate Read Events
+                # for them in our code (`from_query`=True). For such Query Events, we delete the original
+                # Read Events coming from Bigquery audit logs and keep only generated ones.
+
+                # Caveats of SQL parsing approach used here:
+                # 1. If query parsing fails, usage for such query is not considered/counted.
+                # 2. Due to limitations of query parsing, field level usage is not available.
+                # To limit the impact, we use query parsing only for those queries that reference at least
+                # one view. For all other queries, field level usage is available through bigquery audit logs.
+                if (
+                    audit_event.query_event
+                    and audit_event.query_event.referencedViews
+                    and not self.config.usage.apply_view_usage_to_tables
+                ):
+                    query_event = audit_event.query_event
+                    query_event.query_on_view = True
```

Comment:
Could we mark `query_event.query_on_view` when creating the query event rather than here? Doesn't make a big difference to logic, but seems like setting an attribute should be done in object init if possible

Reply:
moved